articles/make-and-keep-your-tests-run-fast.md

```diff
@@ -0,0 +1,53 @@
+
+_Don't write tests in the first place_
+
+Depending on the technology and programming language, there are ways to prevent certain errors by taking advantage of the compiler/type system or static code analysis. This gets rid of tests which need to verify behavior at runtime that can be checked at compile/build time. A simple example would be using an unsigned integer type instead of a signed integer if negative values are not allowed, but it can and should be applied to more complex scenarios as well.
+
+_Separate, separate, separate_
+
+Be clear about what properties are tested on which level (e.g. unit, integration, system) and keep those levels separated. Having clearly separated tests "horizontally" allows defining different guidelines, constraints and approaches to each level and makes it easy to run them independently. In addition to this "horizontal" separation, tests and their corresponding test subjects should also be separated "vertically" by applying the various techniques of modularization. This enables you to run tests only if the corresponding code has been modified.
+
+## What if people are still complaining?
+
+Even after doing everything you can to make running your test suites fast, it likely won't be enough. Be it the sheer amount of test cases or the requirement for a lot of integration and system level testing, at some point things are going to be "slow". But there are some additional strategies for dealing with that:
+
+_Parallelize_
+
+Run test cases and suites in parallel, as much as possible. Do not shy away from distributing the execution across multiple (virtual) machines. This may require additional design and/or refactoring to make that possible. However, if you already applied all of the points above, you usually should be able to parallelize quite a bit without changing anything.
+
+_Prioritize (and fail early)_
+
+Categorize/annotate all tests based on _how essential they are_ and _how quick they are_. The default should be for a test to be "most essential" and "fastest" and mark every test that deviates from that, e.g., exhausting options, testing more (theoretical) corner cases or more complex and slow workflows. Order all from "most essential" to "least essential", then from "fastest" to "slowest", execute them in that order and stop on the first failed test. While this does not help when every test passes, it helps to stop wasting time and resources when something fails.
```

Comment:
Das tönt interessant. Für den Artikel finde ich es gut, dass keine Technologien erwähnt werden. Aber geradie hier würde mich persönlich interessieren, wie du das machst. Etwas in der Art: https://damsteen.nl/blog/2016/06/06/ordered-tests-with-nunit-mstest-xunit-pt2-mstest ?

Reply:
Ja, so etwas kann man dafür verwenden. In NUnit gibt es dafür beispielsweise auch das `Ordered` Attribut.

In machen Projekten hatten wir dieses Konzept schon mit Kategorien umgesetzt, d.h. einfach einzelne Testfälle oder ganze Testsuiten mit Kategorien versehen (z.B. "SmokeTest", ...) und diese dann im Build entsprechend nacheinander ausgeführt. Das ist quasi die Lösung die eigentlich mit jedem Testframework geht, auch wenn das Framework selbst das nicht direkt unterstützt (die meisten unterstützen so etwas wie Test-Kategorien oder -Tags).

(Tatsächlich ist dieser Punkt einer derer, die ich noch am wenigsten im Projekt umgesetzt hatte)